File: ios/runanywhere-sdks-latest/sdk/runanywhere-kotlin/src/commonMain/kotlin/com/runanywhere/sdk/public/RunAnywhere.kt
/*
 * Copyright 2026 RunAnywhere SDK
 * SPDX-License-Identifier: Apache-2.0
 *
 * Main entry point for the RunAnywhere SDK.
 * Thin wrapper that delegates to CppBridge for C++ interop.
 *
 * This file mirrors the iOS SDK's RunAnywhere.swift structure.
 */

package com.runanywhere.sdk.public

import com.runanywhere.sdk.foundation.LogLevel
import com.runanywhere.sdk.foundation.SDKLogger
import com.runanywhere.sdk.public.events.EventBus
import com.runanywhere.sdk.utils.SDKConstants

// ═══════════════════════════════════════════════════════════════════════════
// SDK INITIALIZATION FLOW (Two-Phase Pattern)
// ═══════════════════════════════════════════════════════════════════════════
//
// PHASE 1: Core Init (Synchronous, ~1-5ms, No Network)
// ─────────────────────────────────────────────────────
//   RunAnywhere.initialize(environment)
//     ├─ CppBridge.initialize()
//     │    ├─ PlatformAdapter.register()  ← File ops, logging, keychain
//     │    ├─ Events.register()           ← Analytics callback
//     │    ├─ Telemetry.initialize()      ← HTTP callback
//     │    └─ Device.register()           ← Device registration
//     └─ Mark: isInitialized = true
//
// PHASE 2: Services Init (Async, ~100-500ms, Network May Be Required)
// ────────────────────────────────────────────────────────────────────
//   RunAnywhere.completeServicesInitialization()
//     ├─ CppBridge.initializeServices()
//     │    ├─ ModelAssignment.register()  ← Model assignment callbacks
//     │    └─ Platform.register()         ← LLM/TTS service callbacks
//     └─ Mark: areServicesReady = true
//
// ═══════════════════════════════════════════════════════════════════════════

/**
 * SDK environment configuration.
 */
enum class SDKEnvironment(
    val cEnvironment: Int,
) {
    DEVELOPMENT(0),
    STAGING(1),
    PRODUCTION(2),
    ;

    companion object {
        fun fromCEnvironment(cEnvironment: Int): SDKEnvironment =
            entries.find { it.cEnvironment == cEnvironment } ?: DEVELOPMENT
    }
}

/**
 * The RunAnywhere SDK - Single entry point for on-device AI
 *
 * This object mirrors the iOS RunAnywhere enum pattern, providing:
 * - SDK initialization (two-phase: fast sync + async services)
 * - State access (isInitialized, areServicesReady, version, environment)
 * - Event access via `events` property
 *
 * Feature-specific APIs are available through extension functions in public/extensions/:
 * - STT: RunAnywhere.transcribe(), RunAnywhere.loadSTTModel()
 * - TTS: RunAnywhere.synthesize(), RunAnywhere.loadTTSVoice()
 * - LLM: RunAnywhere.chat(), RunAnywhere.generate(), RunAnywhere.generateStream()
 * - VAD: RunAnywhere.detectSpeech()
 * - VoiceAgent: RunAnywhere.startVoiceSession()
 *
 * All AI component logic (LLM, STT, TTS, VAD) is delegated to the C++ runanywhere-commons
 * layer via CppBridge. Kotlin only handles platform-specific operations (HTTP, audio, file I/O).
 */
object RunAnywhere {
    // ═══════════════════════════════════════════════════════════════════════════
    // MARK: - Private State
    // ═══════════════════════════════════════════════════════════════════════════

    private val logger = SDKLogger("RunAnywhere")

    @Volatile
    private var _currentEnvironment: SDKEnvironment? = null

    @Volatile
    private var _isInitialized: Boolean = false

    @Volatile
    private var _areServicesReady: Boolean = false

    private val lock = Any()

    // ═══════════════════════════════════════════════════════════════════════════
    // MARK: - Public Properties
    // ═══════════════════════════════════════════════════════════════════════════

    /**
     * Check if SDK is initialized (Phase 1 complete)
     */
    val isInitialized: Boolean
        get() = _isInitialized

    /**
     * Alias for isInitialized for compatibility
     */
    val isSDKInitialized: Boolean
        get() = _isInitialized

    /**
     * Check if services are fully ready (Phase 2 complete)
     */
    val areServicesReady: Boolean
        get() = _areServicesReady

    /**
     * Check if SDK is active and ready for use
     */
    val isActive: Boolean
        get() = _isInitialized && _currentEnvironment != null

    /**
     * Current SDK version
     */
    val version: String
        get() = SDKConstants.SDK_VERSION

    /**
     * Current environment (null if not initialized)
     */
    val environment: SDKEnvironment?
        get() = _currentEnvironment

    // ═══════════════════════════════════════════════════════════════════════════
    // MARK: - Event Access
    // ═══════════════════════════════════════════════════════════════════════════

    /**
     * Event bus for SDK event subscriptions.
     *
     * Example usage:
     * ```kotlin
     * RunAnywhere.events.llmEvents.collect { event ->
     *     println("LLM event: ${event.type}")
     * }
     * ```
     */
    val events: EventBus
        get() = EventBus

    // ═══════════════════════════════════════════════════════════════════════════
    // MARK: - Phase 1: Core Initialization (Synchronous)
    // ═══════════════════════════════════════════════════════════════════════════

    /**
     * Initialize the RunAnywhere SDK (Phase 1)
     *
     * This performs fast synchronous initialization. Services initialization
     * is done separately via [completeServicesInitialization].
     *
     * ## Usage Examples
     *
     * ```kotlin
     * // Development mode (default)
     * RunAnywhere.initialize()
     *
     * // Production mode
     * RunAnywhere.initialize(environment = SDKEnvironment.PRODUCTION)
     * ```
     *
     * @param apiKey API key (optional for development, required for production/staging)
     * @param baseURL Backend API base URL (optional)
     * @param environment SDK environment (default: DEVELOPMENT)
     */
    fun initialize(
        apiKey: String? = null,
        baseURL: String? = null,
        environment: SDKEnvironment = SDKEnvironment.DEVELOPMENT,
    ) {
        synchronized(lock) {
            if (_isInitialized) {
                logger.info("SDK already initialized")
                return
            }

            val initStartTime = System.currentTimeMillis()

            try {
                // Store environment
                _currentEnvironment = environment

                // Set log level based on environment
                val logLevel =
                    when (environment) {
                        SDKEnvironment.DEVELOPMENT -> LogLevel.DEBUG
                        SDKEnvironment.STAGING -> LogLevel.INFO
                        SDKEnvironment.PRODUCTION -> LogLevel.WARNING
                    }
                SDKLogger.setLevel(logLevel)

                // Initialize CppBridge (Phase 1)
                // Note: CppBridge is in jvmAndroidMain, we call it via expect/actual
                initializeCppBridge(environment, apiKey, baseURL)

                // Mark Phase 1 complete
                _isInitialized = true

                val initDurationMs = System.currentTimeMillis() - initStartTime
                logger.info("✅ Phase 1 complete in ${initDurationMs}ms (${environment.name})")
            } catch (error: Exception) {
                logger.error("❌ Initialization failed: ${error.message}")
                _currentEnvironment = null
                _isInitialized = false
                throw error
            }
        }
    }

    /**
     * Initialize SDK for development mode (convenience method)
     */
    fun initializeForDevelopment(apiKey: String? = null) {
        initialize(apiKey = apiKey, environment = SDKEnvironment.DEVELOPMENT)
    }

    // ═══════════════════════════════════════════════════════════════════════════
    // MARK: - Phase 2: Services Initialization (Async)
    // ═══════════════════════════════════════════════════════════════════════════

    /**
     * Complete services initialization (Phase 2)
     *
     * Called automatically on first API call, or can be awaited directly.
     * Safe to call multiple times - returns immediately if already done.
     */
    suspend fun completeServicesInitialization() {
        // Fast path: already completed
        if (_areServicesReady) {
            return
        }

        synchronized(lock) {
            if (_areServicesReady) {
                return
            }

            if (!_isInitialized) {
                throw IllegalStateException("SDK must be initialized before completing services initialization")
            }

            logger.info("Initializing services for ${_currentEnvironment?.name} mode...")

            try {
                // Initialize CppBridge services (Phase 2)
                initializeCppBridgeServices()

                // Mark Phase 2 complete
                _areServicesReady = true

                logger.info("✅ Services initialized for ${_currentEnvironment?.name} mode")
            } catch (e: Exception) {
                logger.error("Services initialization failed: ${e.message}")
                throw e
            }
        }
    }

    /**
     * Ensure services are ready before API calls (internal guard)
     * O(1) after first successful initialization
     */
    internal suspend fun ensureServicesReady() {
        if (_areServicesReady) {
            return // O(1) fast path
        }
        completeServicesInitialization()
    }

    /**
     * Ensure SDK is initialized (throws if not)
     */
    internal fun requireInitialized() {
        if (!_isInitialized) {
            throw IllegalStateException("SDK not initialized. Call RunAnywhere.initialize() first.")
        }
    }

    // ═══════════════════════════════════════════════════════════════════════════
    // MARK: - SDK Reset
    // ═══════════════════════════════════════════════════════════════════════════

    /**
     * Reset SDK state
     * Clears all initialization state and releases resources
     */
    suspend fun reset() {
        logger.info("Resetting SDK state...")

        synchronized(lock) {
            // Shutdown CppBridge
            shutdownCppBridge()

            _isInitialized = false
            _areServicesReady = false
            _currentEnvironment = null
        }

        logger.info("SDK state reset completed")
    }

    /**
     * Cleanup SDK resources without full reset
     */
    suspend fun cleanup() {
        logger.info("Cleaning up SDK resources...")
        // Cleanup logic here
    }

    // ═══════════════════════════════════════════════════════════════════════════
    // MARK: - CppBridge Integration (expect/actual pattern)
    // ═══════════════════════════════════════════════════════════════════════════

    /**
     * Initialize CppBridge (Phase 1)
     * Implementation is in jvmAndroidMain via expect/actual
     */
    private fun initializeCppBridge(environment: SDKEnvironment, apiKey: String?, baseURL: String?) {
        logger.debug("CppBridge initialization requested for $environment")
        initializePlatformBridge(environment, apiKey, baseURL)
    }

    /**
     * Initialize CppBridge services (Phase 2)
     * Implementation is in jvmAndroidMain via expect/actual
     */
    private fun initializeCppBridgeServices() {
        logger.debug("CppBridge services initialization requested")
        initializePlatformBridgeServices()
    }

    /**
     * Shutdown CppBridge
     * Implementation is in jvmAndroidMain via expect/actual
     */
    private fun shutdownCppBridge() {
        logger.debug("CppBridge shutdown requested")
        shutdownPlatformBridge()
    }
}

// ═══════════════════════════════════════════════════════════════════════════
// Platform-specific bridge functions (expect/actual pattern)
// ═══════════════════════════════════════════════════════════════════════════

/**
 * Initialize platform-specific bridge (Phase 1).
 * On JVM/Android, this calls CppBridge.initialize() to load native libraries.
 *
 * @param environment SDK environment
 * @param apiKey API key for authentication (required for production/staging)
 * @param baseURL Backend API base URL (required for production/staging)
 */
internal expect fun initializePlatformBridge(environment: SDKEnvironment, apiKey: String?, baseURL: String?)

/**
 * Initialize platform-specific bridge services (Phase 2).
 * On JVM/Android, this calls CppBridge.initializeServices().
 */
internal expect fun initializePlatformBridgeServices()

/**
 * Shutdown platform-specific bridge.
 * On JVM/Android, this calls CppBridge.shutdown().
 */
internal expect fun shutdownPlatformBridge()

File: ios/runanywhere-sdks-latest/sdk/runanywhere-swift/Sources/RunAnywhere/Public/RunAnywhere.swift
//
//  RunAnywhere.swift
//  RunAnywhere SDK
//
//  The main entry point for the RunAnywhere SDK.
//  Contains SDK initialization, state management, and event access.
//

import Combine
import Foundation
#if os(iOS) || os(tvOS) || os(watchOS)
import UIKit
#endif

// MARK: - SDK Initialization Flow
//
// ┌─────────────────────────────────────────────────────────────────────────────┐
// │                         SDK INITIALIZATION FLOW                              │
// └─────────────────────────────────────────────────────────────────────────────┘
//
// PHASE 1: Core Init (Synchronous, ~1-5ms, No Network)
// ─────────────────────────────────────────────────────
//   initialize() or initializeForDevelopment()
//     ├─ Validate params (API key, URL, environment)
//     ├─ Set log level
//     ├─ Store params locally
//     ├─ Store in Keychain (production/staging only)
//     └─ Mark: isInitialized = true
//
// PHASE 2: Services Init (Async, ~100-500ms, Network Required)
// ────────────────────────────────────────────────────────────
//   completeServicesInitialization()
//     ├─ Setup API Client
//     │    ├─ Development: Use Supabase
//     │    └─ Production/Staging: Authenticate with backend
//     ├─ Register C++ Bridge Callbacks
//     │    ├─ Model Assignment (CppBridge.ModelAssignment)
//     │    └─ Platform Services (CppBridge.Platform)
//     ├─ Load Models (from remote API via C++)
//     ├─ Initialize EventPublisher (telemetry → backend)
//     └─ Register Device with Backend
//
// USAGE:
// ──────
//   // Development mode (default)
//   try RunAnywhere.initialize()
//
//   // Production mode - requires API key and backend URL
//   try RunAnywhere.initialize(
//       apiKey: "your_api_key",
//       baseURL: "https://api.runanywhere.ai",
//       environment: .production
//   )
//

/// The RunAnywhere SDK - Single entry point for on-device AI
public enum RunAnywhere {

    // MARK: - Internal State Management

    /// Internal init params storage
    internal static var initParams: SDKInitParams?
    internal static var currentEnvironment: SDKEnvironment?
    internal static var isInitialized = false

    /// Track if services initialization is complete (makes API calls O(1) after first use)
    internal static var hasCompletedServicesInit = false
    /// Track if HTTP/auth setup succeeded (separate from core services so auth can be retried on reconnect)
    internal static var hasCompletedHTTPSetup = false

    // MARK: - SDK State

    /// Check if SDK is initialized (Phase 1 complete)
    public static var isSDKInitialized: Bool {
        isInitialized
    }

    /// Check if services are fully ready (Phase 2 complete)
    public static var areServicesReady: Bool {
        hasCompletedServicesInit
    }

    /// Check if SDK is active and ready for use
    public static var isActive: Bool {
        isInitialized && initParams != nil
    }

    /// Current SDK version
    public static var version: String {
        SDKConstants.version
    }

    /// Current environment (nil if not initialized)
    public static var environment: SDKEnvironment? {
        currentEnvironment
    }

    /// Device ID (Keychain-persisted, survives reinstalls)
    public static var deviceId: String {
        DeviceIdentity.persistentUUID
    }

    // MARK: - Event Access

    /// Access to all SDK events for subscription-based patterns
    public static var events: EventBus {
        EventBus.shared
    }

    // MARK: - Authentication Info (Production/Staging only)

    /// Get current user ID from authentication
    /// - Returns: User ID if authenticated, nil otherwise
    public static func getUserId() -> String? {
        CppBridge.State.userId
    }

    /// Get current organization ID from authentication
    /// - Returns: Organization ID if authenticated, nil otherwise
    public static func getOrganizationId() -> String? {
        CppBridge.State.organizationId
    }

    /// Check if currently authenticated
    /// - Returns: true if authenticated with valid token
    public static var isAuthenticated: Bool {
        CppBridge.State.isAuthenticated
    }

    /// Check if device is registered with backend
    public static func isDeviceRegistered() -> Bool {
        CppBridge.Device.isRegistered
    }

    // MARK: - SDK Reset (Testing)

    /// Reset SDK state (for testing purposes)
    /// Clears all initialization state and cached data
    public static func reset() {
        let logger = SDKLogger(category: "RunAnywhere.Reset")
        logger.info("Resetting SDK state...")

        isInitialized = false
        hasCompletedServicesInit = false
        hasCompletedHTTPSetup = false
        initParams = nil
        currentEnvironment = nil

        // Shutdown all C++ bridges and state
        CppBridge.shutdown()
        CppBridge.State.shutdown()

        logger.info("SDK state reset completed")
    }

    // ═══════════════════════════════════════════════════════════════════════════
    // MARK: - SDK Initialization
    // ═══════════════════════════════════════════════════════════════════════════

    /**
     * Initialize the RunAnywhere SDK
     *
     * This performs fast synchronous initialization, then starts async services in background.
     * The SDK is usable immediately - services will be ready when first API call is made.
     *
     * **Phase 1 (Sync, ~1-5ms):** Validates params, sets up logging, stores config
     * **Phase 2 (Background):** Network auth, service creation, model loading, device registration
     *
     * ## Usage Examples
     *
     * ```swift
     * // Development mode (default)
     * try RunAnywhere.initialize()
     *
     * // Production mode - requires API key and backend URL
     * try RunAnywhere.initialize(
     *     apiKey: "your_api_key",
     *     baseURL: "https://api.runanywhere.ai",
     *     environment: .production
     * )
     * ```
     *
     * - Parameters:
     *   - apiKey: API key (optional for development, required for production/staging)
     *   - baseURL: Backend API base URL (optional for development, required for production/staging)
     *   - environment: SDK environment (default: .development)
     *
     * - Throws: SDKError if validation fails
     */
    public static func initialize(
        apiKey: String? = nil,
        baseURL: String? = nil,
        environment: SDKEnvironment = .development
    ) throws {
        let params: SDKInitParams

        if environment == .development {
            // Development mode - use Supabase, no auth needed
            params = SDKInitParams(forDevelopmentWithAPIKey: apiKey ?? "")
        } else {
            // Production/Staging mode - require API key and URL
            guard let apiKey = apiKey, !apiKey.isEmpty else {
                throw SDKError.general(.invalidConfiguration, "API key is required for \(environment.description) mode")
            }
            guard let baseURL = baseURL, !baseURL.isEmpty else {
                throw SDKError.general(.invalidConfiguration, "Base URL is required for \(environment.description) mode")
            }
            params = try SDKInitParams(apiKey: apiKey, baseURL: baseURL, environment: environment)
        }

        try performCoreInit(with: params, startBackgroundServices: true)
    }

    /// Initialize with URL type for base URL
    public static func initialize(
        apiKey: String,
        baseURL: URL,
        environment: SDKEnvironment = .production
    ) throws {
        let params = try SDKInitParams(apiKey: apiKey, baseURL: baseURL, environment: environment)
        try performCoreInit(with: params, startBackgroundServices: true)
    }

    // ═══════════════════════════════════════════════════════════════════════════
    // MARK: - Phase 1: Core Initialization (Synchronous)
    // ═══════════════════════════════════════════════════════════════════════════

    /// Perform core initialization (Phase 1)
    /// - Parameters:
    ///   - params: SDK initialization parameters
    ///   - startBackgroundServices: If true, starts Phase 2 in background task
    private static func performCoreInit(with params: SDKInitParams, startBackgroundServices: Bool) throws {
        // Return early if already initialized
        guard !isInitialized else { return }

        let initStartTime = CFAbsoluteTimeGetCurrent()

        // Step 1: Set environment FIRST so Logging.shared initializes with correct config
        // This must happen before any SDKLogger usage to ensure logs appear correctly
        currentEnvironment = params.environment
        initParams = params

        // Step 2: Apply environment-specific logging configuration
        Logging.shared.applyEnvironmentConfiguration(params.environment)

        // Step 3: Initialize all core C++ bridges (platform adapter, events, telemetry, device)
        // This must happen early so all C++ logs route to SDKLogger and events can be emitted
        CppBridge.initialize(environment: params.environment)

        // Now safe to create logger and track events
        let logger = SDKLogger(category: "RunAnywhere.Init")
        CppBridge.Events.emitSDKInitStarted()

        do {

            // Step 4: Persist to Keychain (production/staging only)
            if params.environment != .development {
                try KeychainManager.shared.storeSDKParams(params)
            }

            // Mark Phase 1 complete
            isInitialized = true

            let initDurationMs = (CFAbsoluteTimeGetCurrent() - initStartTime) * 1000
            logger.info("✅ Phase 1 complete in \(String(format: "%.1f", initDurationMs))ms (\(params.environment.description))")

            CppBridge.Events.emitSDKInitCompleted(durationMs: initDurationMs)

            // Optionally start Phase 2 in background
            if startBackgroundServices {
                logger.debug("Starting Phase 2 (services) in background...")
                Task.detached(priority: .userInitiated) {
                    do {
                        try await completeServicesInitialization()
                        SDKLogger(category: "RunAnywhere.Init").info("✅ Phase 2 complete (background)")
                    } catch {
                        SDKLogger(category: "RunAnywhere.Init")
                            .warning("⚠️ Phase 2 failed (non-critical): \(error.localizedDescription)")
                    }
                }
            }

        } catch {
            logger.error("❌ Initialization failed: \(error.localizedDescription)")
            initParams = nil
            isInitialized = false
            CppBridge.Events.emitSDKInitFailed(error: SDKError.from(error))
            throw error
        }
    }

    // ═══════════════════════════════════════════════════════════════════════════
    // MARK: - Phase 2: Services Initialization (Async)
    // ═══════════════════════════════════════════════════════════════════════════

    /// Complete services initialization (Phase 2)
    ///
    /// Called automatically in background by `initialize()`, or can be awaited directly
    /// via `initializeAsync()`. Safe to call multiple times — returns immediately if Phase 2
    /// is already complete. Note: if initialization succeeded in offline mode (HTTP/auth setup
    /// failed), this fast-path still returns immediately. HTTP/auth retry is handled
    /// automatically by `ensureServicesReady()` on the next API call.
    ///
    /// This method:
    /// 1. Sets up API client (with authentication for production/staging)
    /// 2. Initializes C++ model registry and bridges
    /// 3. Initializes EventPublisher for telemetry
    /// 4. Registers device with backend
    public static func completeServicesInitialization() async throws {
        // Fast path: already completed
        if hasCompletedServicesInit {
            return
        }

        guard let params = initParams, let environment = currentEnvironment else {
            throw SDKError.general(.notInitialized, "SDK not initialized")
        }

        let logger = SDKLogger(category: "RunAnywhere.Services")

        // Check if HTTP needs initialization
        let httpNeedsInit = await !CppBridge.HTTP.shared.isConfigured

        if httpNeedsInit {
            logger.info("Initializing services for \(environment.description) mode...")

            // Step 1: Configure HTTP transport
            do {
                try await setupHTTP(params: params, environment: environment, logger: logger)
                hasCompletedHTTPSetup = true
            } catch {
                // If HTTP/auth setup fails (e.g. device is offline), log warning but
                // continue initialization so local/cached models remain accessible.
                logger.warning("⚠️ HTTP/Auth setup failed (offline?): \(error.localizedDescription)")
                logger.info("Continuing SDK init in offline mode – local models will be available")
            }

            // Step 1.5: Flush any queued telemetry events (may be no-op if HTTP unconfigured)
            CppBridge.Telemetry.flush()
            logger.debug("Attempted telemetry flush (may be no-op if HTTP unconfigured)")
        }

        // Step 2: Initialize C++ state
        CppBridge.State.initialize(
            environment: environment,
            apiKey: params.apiKey,
            baseURL: params.baseURL,
            deviceId: DeviceIdentity.persistentUUID
        )
        logger.debug("C++ state initialized")

        // Step 3: Initialize service bridges (Platform, ModelAssignment)
        // Must be on MainActor for Platform services
        await MainActor.run {
            CppBridge.initializeServices()
        }
        logger.debug("Service bridges initialized")

        // Step 4: Set base directory for C++ model paths
        if let documentsURL = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first {
            try CppBridge.ModelPaths.setBaseDirectory(documentsURL)
            logger.debug("Model paths base directory set")
        }

        // Step 5: Register device via CppBridge (C++ handles all business logic)
        do {
            try await CppBridge.Device.registerIfNeeded(environment: environment)
            logger.debug("Device registration check completed")
        } catch {
            logger.warning("Device registration failed (non-critical): \(error.localizedDescription)")
        }

        // Step 6: Discover already-downloaded models on file system
        // This scans the models directory and updates the registry for models found on disk
        let discoveryResult = await CppBridge.ModelRegistry.shared.discoverDownloadedModels()
        if discoveryResult.discoveredCount > 0 {
            logger.info("Discovered \(discoveryResult.discoveredCount) downloaded models on startup")
        }

        // Mark Phase 2 complete
        hasCompletedServicesInit = true
    }

    /// Ensure services are ready before API calls (internal guard)
    /// O(1) after first successful initialization with HTTP configured.
    /// If core services are done but HTTP/auth failed (offline init), retries auth only.
    internal static func ensureServicesReady() async throws {
        if hasCompletedServicesInit && hasCompletedHTTPSetup {
            return // O(1) fast path — fully initialized
        }
        if hasCompletedServicesInit && !hasCompletedHTTPSetup {
            // Core services done, but HTTP/auth failed earlier (offline init).
            // Retry HTTP setup only — safe because setupHTTP is idempotent.
            await retryHTTPSetup()
            return
        }
        try await completeServicesInitialization()
    }

    /// Retry HTTP/auth setup after an offline initialization.
    /// Safe to call multiple times — checks `CppBridge.HTTP.shared.isConfigured` first.
    /// Failures are silently logged; the next `ensureServicesReady()` call will retry.
    private static func retryHTTPSetup() async {
        guard let params = initParams, let environment = currentEnvironment else { return }
        let logger = SDKLogger(category: "RunAnywhere.HTTPRetry")

        let httpNeedsInit = await !CppBridge.HTTP.shared.isConfigured
        guard httpNeedsInit else {
            hasCompletedHTTPSetup = true
            return
        }

        do {
            try await setupHTTP(params: params, environment: environment, logger: logger)
            hasCompletedHTTPSetup = true
            logger.info("✅ HTTP/Auth setup succeeded on retry")

            // Flush any telemetry events queued during offline period
            CppBridge.Telemetry.flush()
            logger.debug("Flushed queued telemetry after successful HTTP retry")
        } catch {
            logger.debug("HTTP/Auth retry failed (still offline?): \(error.localizedDescription)")
        }
    }

    // ═══════════════════════════════════════════════════════════════════════════
    // MARK: - Private: Service Setup Helpers
    // ═══════════════════════════════════════════════════════════════════════════

    /// Setup HTTP transport via CppBridge.HTTP
    private static func setupHTTP(
        params: SDKInitParams,
        environment: SDKEnvironment,
        logger: SDKLogger
    ) async throws {
        switch environment {
        case .development:
            // Use C++ development config for Supabase (cross-platform)
            if await CppBridge.DevConfig.configureHTTP() {
                logger.debug("HTTP: Supabase from C++ config (development)")
            } else {
                await CppBridge.HTTP.shared.configure(baseURL: params.baseURL, apiKey: params.apiKey)
                logger.debug("HTTP: Provided URL (development)")
            }

        case .staging, .production:
            // Configure HTTP first
            await CppBridge.HTTP.shared.configure(baseURL: params.baseURL, apiKey: params.apiKey)

            // Authenticate via CppBridge.Auth
            try await CppBridge.Auth.authenticate(apiKey: params.apiKey)
            logger.info("Authenticated for \(environment.description)")
        }
    }

}
